packages/circuits/crates/libs/polynomial/src/lib.nr

```diff
@@ -98,38 +98,87 @@ impl<let N: u32> Polynomial<N> {
-/// ordering of coefficients to ensure reproducible challenge values.
-///
-/// # Generic Parameters
-/// * `A` - The degree of each polynomial in the array
-/// * `L` - The number of polynomials in the array
-/// * `SIZE` - The total size of the target array
-///
-/// # Arguments
-/// * `inputs` - The target array to store flattened coefficients
-/// * `poly` - Array of polynomials to flatten
-/// * `offset` - Starting position in the inputs array
-///
-/// # Returns
-/// A tuple containing:
-/// * The updated inputs array with flattened coefficients
-/// * The new offset position after flattening
-pub fn flatten<let A: u32, let L: u32, let SIZE: u32>(
-    mut inputs: [Field; SIZE],
+/// Flatten `L` polynomials (each `Polynomial<A>`) into a linear Vec<Field> of packed carriers.
+/// - Uses the same hex-aligned packing as `packer::<A, BIT>`:
+///   `nibble_bits = ceil_to_nibble(BIT)`, `limb_bits = nibble_bits + 4`,
+///   `group = 254 / limb_bits` carriers per chunk.
+/// - Preserves deterministic order: iterate polynomials in `j`, then append all their carriers.
+pub fn flatten<let A: u32, let L: u32, let BIT: u32>(
+    mut inputs: Vec<Field>,
     poly: [Polynomial<A>; L],
-    mut offset: u32,
-) -> ([Field; SIZE], u32) {
+) -> Vec<Field> {
+    // Compute packing layout (must mirror `packer`).
+    let nibble_bits = ((BIT + 3) / 4) * 4; // ceil to multiple of 4
+    let limb_bits = nibble_bits + 4; // +1 nibble offset
+    assert(limb_bits <= 254);
+
+    let group = 254 / limb_bits; // max limbs per field
+    assert(group >= 1);
+
+    // Carriers produced per polynomial.
+    let num_chunks = (A + group - 1) / group;
+
+    // Pack each polynomial, then append all its carriers in order.
     for j in 0..L {
-        for i in 0..A {
-            inputs[i + offset] = poly[j].coefficients[i];
+        let packed = packer::<A, BIT>(poly[j].coefficients);
+        // `packed.len()` == `num_chunks`; push explicitly to be clear/stable.
+        for i in 0..num_chunks {
+            inputs.push(packed.get(i));
         }
-        offset += A;
     }
 
-    (inputs, offset)
+    inputs
+}
+
+/// Pack `values` into as few `Field` carriers as possible (hex-aligned).
+/// - `BIT`: bit width of each value (before offset). We round it up to a nibble.
+/// - Limb = `v + 2^nibble_bits` (lifts negatives); radix per limb = `2^(nibble_bits+4)`.
+/// - Packs `group = 254 / (nibble_bits+4)` values per carrier, big-endian; pads tail with `2^nibble_bits`.
+pub fn packer<let A: u32, let BIT: u32>(values: [Field; A]) -> Vec<Field> {
+    assert(BIT >= 1);
+
+    // Align to hex digits (4 bits) and add +1 nibble headroom.
+    let nibble_bits = ((BIT + 3) / 4) * 4; // ceil_to_nibble(BIT)
+    let limb_bits = nibble_bits + 4; // limb width used for packing
+    assert(limb_bits <= 254);
+
+    // Max limbs per 254-bit field element.
+    let group = 254 / limb_bits;
+    assert(group >= 1);
```

Comment:
these lines are repeated in L110-L116. can we extract them in a meaningful and reusable function?